### PR TITLE
Clipping

### PR DESCRIPTION
**Pull request by royess**
As described in #58 , the PR includes clipping algorithm, bigram, bipartite entanglement.

### CODE REVIEW

**Comment by Krastanov** [comment]
Thanks! I will do a quick first pass review, but there is no need to start addressing these comments until a full review is done.

File: src/entanglement.jl [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+    lowbit = Tme(0x1)
+    zerobit = Tme(0x0)
+    rows, columns = size(stabilizerview(state))
+    # step 1: pregauge
+    i = 1 # index to place used stab
+    for j in 1:columns
+        jbig = _div(Tme,j-1)+1
+        jsmall = lowbit<<_mod(Tme,j-1)
+        # find first row that is not I in col j
+        k1 = findfirst(e->e&jsmall!=zerobit, xs[jbig,i:end] .| zs[jbig,i:end])
```
[the remark this answers is not in the document]
Comment (Krastanov) [reply src/entanglement.jl:18]: In terms of allocations, I performed the following test

```
s = random_stabilizer(10)
@benchmark clip!(_s) setup=(_s=copy(s)) evals=1
@benchmark canonicalize!(_s) setup=(_s=copy(s)) evals=1
```

`canonicalize!` and `clip!` are both some forms of Gaussian elimination, and while `clip!` is certainly much more sophisticated, I think it is reasonable to try to get them within the same order of magnitude. Hopefully it would make `clip!` about 5 times faster. Anyway, let's not worry about the non-obvious optimizations for the moment and just focus on things like the more obvious allocations like the one above.

**Review by Krastanov — COMMENTED** [review]
Here are some initial first impressions. I think it is fairly close to ready, I just want to read the referenced papers in a bit more detail and I would like us to discuss the points I had raised.

Please also do the following:
- Add the references to the bibtex file in the documentation folder
- If you are comfortable with it, add your name and email to the CONTRIBUTORS file

File: src/graphs.jl [hunk 2]
```diff
@@ -176,3 +176,15 @@ function graph_gate(h_idx, ip_idx, z_idx, n)
     end
     c
 end
+
+
+"""
+Get bipartite entanglement by first convertig the state to a graph.
+"""
+function entanglement_from_graph(state::Stabilizer, subsystem)
```
[the remark this answers is not in the document]
Comment (Krastanov) [reply src/graphs.jl:184]: one minor thing to add to the graph implementation. There is a significant allocation done in an inner loop (the use of `collect`). I would suggest doing the following
`other_subsistem = setdiff(1:nv(graph), subsystem)`

File: test/test_entanglement.jl [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+                    @debug s
+                    graph = Graphs.Graph(s)
+                    @debug collect(Graphs.edges(graph))
+                end
+            end  
+        end
+    end
+end
+
+function test_entanglement_from_graph()
```
Comment (Krastanov) [inline test/test_entanglement.jl:58]: Could you add the problematic example we discussed in the other thread as an explicit separete test?

https://github.com/Krastanov/QuantumClifford.jl/issues/57#issuecomment-1146259434=

File: test/test_entanglement.jl [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+            end
+        end
+    end
+end
+
+
+# functions to show info when fails
+onfail(body, _::Test.Pass) = nothing
+onfail(body, _::Test.Fail) = body()
+onfail(body, _::Tuple{Test.Fail,T}) where {T} = body()
```
[the remark this answers is not in the document]
Comment (royess) [reply test/test_entanglement.jl:22]: The snippet is from <https://discourse.julialang.org/t/print-debug-info-for-failed-test/22311>, I will add a credit in comments. And I agree that we can move it to the main test running file.

File: test/test_entanglement.jl [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+                    @debug s
+                    graph = Graphs.Graph(s)
+                    @debug collect(Graphs.edges(graph))
+                end
+            end  
+        end
+    end
+end
+
+import HDF5
```
[the remark this answers is not in the document]
Comment (royess) [reply test/test_entanglement.jl:79]: Yes. But I am also wondering whether we should include Qiskit test function in this PR? Because it has non-julia dependence.

File: src/entanglement.jl [hunk 6]
```diff
@@ -0,0 +1,148 @@
+"""
+Get the clipped gauge of a stablizer state. 
+"""
+function clip!(state::AbstractStabilizer; phases::Val{B}=Val(true)) where B
+    xzs = stabilizerview(state).xzs
+    xs = @view xzs[1:end÷2,:]
+    zs = @view xzs[end÷2+1:end,:]
+    Tme = eltype(xzs)
+    lowbit = Tme(0x1)
+    zerobit = Tme(0x0)
```
[the remark this answers is not in the document]
Comment (royess) [reply src/entanglement.jl:10]: As I mentioned in #58 (and previously wrongly referred as `biset`), I wonder whether we can use `biarray` as an abstract, which is offered by `Base.jl`. (<https://github.com/JuliaLang/julia/blob/master/base/bitarray.jl>)

File: src/graphs.jl (hunk 2 above)
[the remark this answers is not in the document]
Comment (royess) [reply src/graphs.jl:184]: > Maybe we can do something like `entanglement(state, algorithm=FromGraph())` or `entanglement(state, algorithm=Val(:fromgraph))`?

Sure. But here is a tricky thing: entanglement calculation based on clipping algorithm only supports **contiguous subregion**, while calculation based on graph and on mixed state entropy do not have such limitations. For sure we can add an additional permutation to clipping algorithm to support non-contiguous subregion.

Or, I suppose we can give `entanglement` two methods:

1. `entanglement(state, subsystem, algorithm)`. And `algorithm` can be selected from "mixed state", "graph".
2. `entanglement(state, leftendpoint, rightendpoint, algorithm)`.  And `algorithm` can be selected from "mixed state", "clipping" or "graph".

I would like  to hear your opinions on this.

---

> I would suggest doing the following
> `other_subsistem = setdiff(1:nv(graph), subsystem)`

That makes sense. Thanks!

[the remark this answers is not in the document]
Comment (Krastanov) [reply src/graphs.jl:184]: Would the following work instead:
```
function entanglement_entropy(state, range::AbstractRange, ::Val{:clipping})
function entanglement_entropy(state, range::AbstractVector, ::Val{:traceout})
function entanglement_entropy(state, range::AbstractVector, ::Val{:graph})
```

The `clipping` method is defined only on `AbstractRange`. while the other are defined for `AbstractVector`.

File: test/test_entanglement.jl (hunk 5 above)
[the remark this answers is not in the document]
Comment (Krastanov) [reply test/test_entanglement.jl:79]: I think at this point it is not necessary. There are three independent checks now, so adding a fourth one does not have that much value.

File: src/entanglement.jl (hunk 6 above)
[the remark this answers is not in the document]
Comment (Krastanov) [reply src/entanglement.jl:10]: The 2-axis implementation of that array is very inconvenient for any broadcasted operations as it does not "pad" rows, so the matrix [true true; false false] would be stored in a single UInt, making row operations difficult.

But we actually already have all the features of the bitarray - the way setting and getting of indices is implemented the same way and you can do `stab[1,2]` if you want to - you could have written all of your code in that fashion, not using any bitwise operations. But it is not obvious how to ensure the compiler is smart enough to make that higher-level code just as efficient. This would be something to explore.

File: src/graphs.jl (hunk 2 above)
[the remark this answers is not in the document]
Comment (royess) [reply src/graphs.jl:184]: This is a good suggestion! I think we should define `clipping` on `UnitRange` only:
```julia
function entanglement_entropy(state, range::UnitRange, ::Val{:clipping})
```

File: src/entanglement.jl (hunk 6 above)
[the remark this answers is not in the document]
Comment (Krastanov) [reply src/entanglement.jl:10]: It seems the code can indeed be significantly simplified without performance losses. The original canonicalization routines were written before the getindex and setindex, and I never changed them as I did not expect the compiler to be able to optimize them. But the following commit seems to work without a change in performance: https://github.com/Krastanov/QuantumClifford.jl/compare/master...simplecanonicalization

No need to make such changes for the moment, but just wanted to let you know, in case you are making future changes and want something easier to write.

File: src/entanglement.jl [hunk 7]
```diff
@@ -0,0 +1,148 @@
+"""
+Get the clipped gauge of a stablizer state. 
+"""
+function clip!(state::AbstractStabilizer; phases::Val{B}=Val(true)) where B
```
[the remark this answers is not in the document]
Comment (Krastanov) [reply src/entanglement.jl:4]: And let's add a bit to the documentation for it.

In the docstring, could you add "Introduced in [nahum2017quantum](@cite), with a more detailed explanation of the algorithm in Appendix A of [li2019measurement](@cite)" and add the two papers to the bibtex file in `./docs/src`. Some of them might already be in the bibtex file.

Then could you also mention that this canonicalization operation exists in a new section in the `./docs/src/canonicalization.md` documentation file.

File: src/entanglement.jl (hunk 1 above)
[the remark this answers is not in the document]
Comment (Krastanov) [reply src/entanglement.jl:18]: Using the new getters and setters that were implemented historically after the various canonicalization operations, this can be written as `findfirst(ii->|(tab[ii,j]...), i:rows)` and should be faster due to the lack of allocation of temporary arrays.

File: src/entanglement.jl [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+                deleteat!(unfrozen_rows, k1)
+            end
+        end
+    end
+    state
+end
+
+
+"""
+Get bigram which contains the location of endpoints.
```
Comment (Krastanov) [inline src/entanglement.jl:112]: Could you add a reference, same as the one in `canonicalize_clip!`?

Comment (Krastanov) [reply src/entanglement.jl:112]: And a ```See also: [`canonicalize_clip!`](@ref)```

File: test/test_entanglement.jl [hunk 9]
```diff
@@ -0,0 +1,125 @@
+function test_clipping()
+    @testset "Clipped gauge of stabilizer states" begin
+        for n in test_sizes
+            num_repeat = 100
+            for i_repeat in 1:num_repeat
+                s = random_stabilizer(n)
```
Comment (Krastanov) [inline test/test_entanglement.jl:6]: could you also add a test that checks that the canonicalization does not break invariants. Something like

`@test stab_looks_good(canonicalize_clip!(copy(s)))` (that function is already implemented and used in other tests).

Also `@test canonicalize!(canonicalize_clip!(copy(s)))==canonicalize!(copy(s))`

File: test/test_entanglement.jl (hunk 5 above)
[the remark this answers is not in the document]
Comment (Krastanov) [reply test/test_entanglement.jl:79]: Yeah, let's just skip the qiskit tests now that we have multiple different ways to test (but keep the tests between different internal implementations).

**Review by Krastanov — COMMENTED** [review]
The only thing I still need to review is the bigram function. Consider the review of the rest of the functionality complete.

File: src/graphs.jl (hunk 2 above)
[the remark this answers is not in the document]
Comment (Krastanov) [reply src/graphs.jl:184]: yes, UnitRange is indeed a better choice

File: src/entanglement.jl [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+"""
+function get_bigram(state::AbstractStabilizer; do_clip::Bool=true)
```
Comment (Krastanov) [inline src/entanglement.jl:114]: Could we shorten this to just `bigram`? Similar to how other conversion operations in this library just have the name of what they convert to.

**Review by Krastanov — COMMENTED** [review]
A few comments on the `bigram` code.

**Comment by Krastanov** [comment]
I think these are all the comments I have. This is a great addition, thanks for writing it.

File: src/graphs.jl (hunk 2 above)
[the remark this answers is not in the document]
Comment (Krastanov) [reply src/graphs.jl:184]: And could you leave a reference to https://arxiv.org/abs/quant-ph/0602096 in the docstring?

File: test/test_entanglement.jl (hunk 5 above)
[the remark this answers is not in the document]
Comment (royess) [reply test/test_entanglement.jl]: Done in https://github.com/Krastanov/QuantumClifford.jl/pull/59/commits/9b0b9cf933e960bd1786f71c3b2077dc70db40d1

File: src/entanglement.jl (hunk 1 above)
[the remark this answers is not in the document]
Comment (royess) [reply src/entanglement.jl:18]: Thanks for your suggestions. I replace `.|` to get:
```julia
julia> @benchmark canonicalize!(_s) setup=(_s=copy(s)) evals=1000
BenchmarkTools.Trial: 7002 samples with 1000 evaluations.
 Range (min … max):  631.366 ns …   4.990 μs  ┊ GC (min … max): 0.00% … 83.25%
 Time  (median):     684.457 ns               ┊ GC (median):    0.00%
 Time  (mean ± σ):   712.691 ns ± 161.646 ns  ┊ GC (mean ± σ):  0.71% ±  3.04%

           ▇█▁
  ▇▄▂▂▃▃█▅▄███▇▆▄▄▂▂▂▂▂▂▂▂▂▂▂▂▃▃▂▂▃▃▂▂▂▂▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁ ▂
  631 ns           Histogram: frequency by time          930 ns <

 Memory estimate: 64 bytes, allocs estimate: 2.

julia> @benchmark clip!(_s) setup=(_s=copy(s)) evals=1000
BenchmarkTools.Trial: 1345 samples with 1000 evaluations.
 Range (min … max):  2.946 μs … 14.066 μs  ┊ GC (min … max):  0.00% … 51.92%
 Time  (median):     3.205 μs              ┊ GC (median):     0.00%
 Time  (mean ± σ):   3.714 μs ±  1.657 μs  ┊ GC (mean ± σ):  11.84% ± 16.78%

  ▃██▃▂▄▁                                                ▁
  ████████▁▄▄▄▁▁▄▄▁▁▁▁▁▁▁▁▁▄▄▁▁▄▄▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▄▆█████▆ █
  2.95 μs      Histogram: log(frequency) by time     9.86 μs <

 Memory estimate: 3.16 KiB, allocs estimate: 56.
```

And before the modification:
```julia
BenchmarkTools.Trial: 972 samples with 1000 evaluations.
 Range (min … max):  3.746 μs … 12.124 μs  ┊ GC (min … max):  0.00% … 50.60%
 Time  (median):     4.100 μs              ┊ GC (median):     0.00%
 Time  (mean ± σ):   5.142 μs ±  2.179 μs  ┊ GC (mean ± σ):  19.33% ± 21.86%

  ▁▆█▇▃▂▂▂                                       ▂▁▁▂▁ ▁
  █████████▅▁▅▁▄▅▅▅▅▅▅▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▅▇████████▄▆▆ █
  3.75 μs      Histogram: log(frequency) by time     10.6 μs <

 Memory estimate: 8.06 KiB, allocs estimate: 112.
```

I suppose there is indeed more optimization to do, including getting rid of memory allocs.

File: src/entanglement.jl [hunk 11]
```diff
@@ -0,0 +1,148 @@
...
+    if do_clip
```
[the remark this answers is not in the document]
Comment (royess) [reply src/entanglement.jl]: Thank! Done in https://github.com/Krastanov/QuantumClifford.jl/pull/59/commits/c3f6f0b00c316d4695eee6c1d1204a9cffbc7643.

File: src/entanglement.jl [hunk 12]
```diff
@@ -0,0 +1,148 @@
...
+end
+
+
+"""
+Get bipartite entanglement of a contiguous subsystem.
+"""
+function entanglement_cont(state::AbstractStabilizer, subsystem_ends; do_clip::Bool=true)
+    bigram = get_bigram(state; do_clip=do_clip)
+    leftend, rightend = subsystem_ends
+    0.5 * count(r->(leftend<=r[1]<=rightend)⊻(leftend<=r[2]<=rightend), eachrow(bigram))
```
[the remark this answers is not in the document]
Comment (royess) [reply src/entanglement.jl]: Thank! Done in https://github.com/Krastanov/QuantumClifford.jl/pull/59/commits/c3f6f0b00c316d4695eee6c1d1204a9cffbc7643.

File: src/entanglement.jl [hunk 13]
```diff
@@ -0,0 +1,148 @@
...
+            xorzs[jbig,i]&jsmall!=zerobit), 1:columns)
+    end
+    bigram
+end
+
+
+"""
+Get bipartite entanglement of a contiguous subsystem.
+"""
+function entanglement_cont(state::AbstractStabilizer, subsystem_ends; do_clip::Bool=true)
```
[the remark this answers is not in the document]
Comment (royess) [reply src/entanglement.jl]: Thank! Done in https://github.com/Krastanov/QuantumClifford.jl/pull/59/commits/c3f6f0b00c316d4695eee6c1d1204a9cffbc7643.

File: src/entanglement.jl [hunk 14]
```diff
@@ -0,0 +1,148 @@
...
+    xs = @view xzs[1:end÷2,:]
+    zs = @view xzs[end÷2+1:end,:]
+    Tme = eltype(xzs)
+    lowbit = Tme(0x1)
+    zerobit = Tme(0x0)
+    rows, columns = size(stabilizerview(state))
+    xorzs = xs .| zs
+    bigram = zeros(Int, rows, 2)
+    for i in 1:rows
+        bigram[i, 1] = findfirst(j->(
```
[the remark this answers is not in the document]
Comment (royess) [reply src/entanglement.jl]: In my codes, I choose to first calculate `xorzs` and store it as a temp array, which is re-used in every iteration over rows. I suppose it makes a better performance for a large number of qubits. And the simplification you suggest seems to require moving the `|` into the predicate. I would like to hear your opinions on this.

[the remark this answers is not in the document]
Comment (Krastanov) [reply src/entanglement.jl]: I do not think one can answer which one is better without benchmarking. My intuition is that creating the large temporary array would be slower than simply moving the calculation to the predicate, but that is just a guess. Feel free to leave it as it is, and maybe just put in a "TODO check whether `findfirst(j->|(tab[i,j]...), 1:columns)` is faster".

File: test/test_entanglement.jl (hunk 4 above)
[the remark this answers is not in the document]
Comment (royess) [reply test/test_entanglement.jl]: Done in https://github.com/Krastanov/QuantumClifford.jl/pull/59/commits/b106918a23dd0d754b66e0567efac69797ba9bd3.

File: test/test_entanglement.jl (hunk 3 above)
Comment (royess) [reply test/test_entanglement.jl:54]: Thanks! Done in https://github.com/Krastanov/QuantumClifford.jl/pull/59/commits/b106918a23dd0d754b66e0567efac69797ba9bd3.

File: test/test_entanglement.jl (hunk 9 above)
Comment (royess) [reply test/test_entanglement.jl:6]: Thanks! Done in https://github.com/Krastanov/QuantumClifford.jl/commit/b106918a23dd0d754b66e0567efac69797ba9bd3 and https://github.com/Krastanov/QuantumClifford.jl/pull/59/commits/0730c65bd84de72e1e20767128cbe312ad72438e.

File: src/entanglement.jl (hunk 1 above)
[the remark this answers is not in the document]
Comment (Krastanov) [reply src/entanglement.jl]: Let's just get the current implementation in, we can worry about optimizations later. Just add a TODO comment at the start of the function mentioning that it has spurious allocation.

In terms of how to approach this problem, the slowdown is almost certainly due to whatever is causing the allocations. A first pass of just looking at the code is a good first step, looking for any `collect` calls or anything creating arrays. If that manual check does not help, the second thing to do would be to use the profiler to track allocations (it requires starting julia with a special flag). But again, this is already fast enough, so our first priority should be to merge it in.

[the remark this answers is not in the document]
Comment (Krastanov) [reply src/entanglement.jl]: Also, for tests that modify the entity you are testing, I think it is important to set evals=1. My understanding is limited, and I might be wrong, but I think setup is performed once per "sample", while each sample performs multiple "evals" of the expression. So if you do not want to canonicalize an already canonical tableau, you need evals=1. For more details you can check the BenchmarkTools manual https://juliaci.github.io/BenchmarkTools.jl/dev/manual/

File: src/entanglement.jl [hunk 15]
```diff
@@ -0,0 +1,148 @@
...
+            jbig = _div(Tme,j-1)+1;
+            jsmall = lowbit<<_mod(Tme,j-1);
+            xorzs[jbig,i]&jsmall!=zerobit), 1:columns)
+    end
+    bigram
+end
+
+
+"""
+Get bipartite entanglement of a contiguous subsystem.
```
[the remark this answers is not in the document]
Comment (royess) [reply src/entanglement.jl]: Thanks! Done in https://github.com/Krastanov/QuantumClifford.jl/pull/59/commits/0cf5fe5b017821af29b453f729195ff03df5f750.

File: src/graphs.jl (hunk 2 above)
[the remark this answers is not in the document]
Comment (royess) [reply src/graphs.jl]: Thanks! Done in https://github.com/Krastanov/QuantumClifford.jl/pull/59/commits/0cf5fe5b017821af29b453f729195ff03df5f750.

File: src/entanglement.jl (hunk 7 above)
[the remark this answers is not in the document]
Comment (royess) [reply src/entanglement.jl]: Thanks! Done in https://github.com/Krastanov/QuantumClifford.jl/pull/59/commits/0cf5fe5b017821af29b453f729195ff03df5f750.

**Comment by royess** [comment]
Hi, @Krastanov , thanks again for your detailed reviews! By https://github.com/Krastanov/QuantumClifford.jl/pull/59/commits/b721b98fe52d840960c4690addc4033a4734d332, I suppose all modifications are completed according to our discussion.

**Comment by royess** [comment]
> It seems there is a bibtex reference missing in the docs (check the failing build above). It is `nahum2017quantum`. Could you add it in?

I wrote a wrong ref entry name. Now should be good.

**Comment by Krastanov** [comment]
Great! This is merged now. Thank you for contributing this and let me know if there are other related tools that would be useful to you that we can look into adding!

I will release a new version towards the end of the week: there are some minor addition that I would like to look into first.
